logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
+            this.resolverRegister = new HashMap<>();
+            this.sysVarCounter = 0;
+            this.valueTypeRegister = new HashMap<>();
             conjunction.variables().forEach(this::convert);
         }
 
-        private TypeVariable retrieveTypeFromLabel(Label rootLabel) {
-            Optional<TypeVariable> rootType = iterate(conjunction.variables())
-                    .filter(Variable::isType).map(Variable::asType)
-                    .filter(v -> v.label().isPresent() && v.label().get().properLabel().equals(rootLabel)).first();
-
-            if (rootType.isPresent()) {
-                TypeVariable convertedRootType = convert(rootType.get());
-                neighbours.putIfAbsent(convertedRootType, new HashSet<>());
-                return convertedRootType;
-            }
-
-            TypeVariable newRootType = new TypeVariable(Identifier.Variable.of(Reference.label(rootLabel.toString())));
-            newRootType.label(rootLabel);
-            neighbours.putIfAbsent(newRootType, new HashSet<>());
-            return newRootType;
-        }
-
-        Resolvers resolvers() {
-            return resolvers;
-        }
-
-        HashMap<TypeVariable, Set<TypeVariable>> neighbours() {
-            return neighbours;
+        Traversal resolverTraversal() {
+            return resolverTraversal;
         }
 
-        private boolean hasNoInfo(TypeVariable variable) {
-            return (!variable.reference().isLabel() &&
-                    !variable.sub().isPresent() &&
-                    !variable.label().isPresent() &&
-                    variable.is().isEmpty());
+        Variable getVariable(Reference reference) {
+            return variableRegister.get(reference);
         }
 
-        private TypeVariable convert(Variable variable) {
-            if (variable.isType()) return convert(variable.asType());
-            else if (variable.isThing()) return convert(variable.asThing());
+        private void convert(Variable variable) {
+            if (variable.isType()) convert(variable.asType());
+            else if (variable.isThing()) convert(variable.asThing());
             else throw GraknException.of(ILLEGAL_STATE);
         }
 
         private TypeVariable convert(TypeVariable variable) {
-            if (resolvers.contains(variable)) return resolvers.resolver(variable);
-
-            TypeVariable resolver = resolvers.register(variable);
-            neighbours.putIfAbsent(resolver, new HashSet<>());
-
-            if (variable.constraints().isEmpty()) return resolver;
-
-            resolver = cloneVariable(variable);
-            neighbours.putIfAbsent(resolver, new HashSet<>());
-            return resolver;
+            if (resolverRegister.containsKey(variable.id())) return resolverRegister.get(variable.id());
+            resolverRegister.put(variable.id(), variable);
+            variableRegister.putIfAbsent(variable.reference(), variable);
+            variable.addTo(resolverTraversal);
```

Comment:
In fact, when a `Variable` is already a `TypeVariable`, we have only to add its to the traversal in the ordinary way. That's pretty convinient!

Reply:
This was a very valid with the previous iteration of `TypeResolver`. But we are no longer manipulating constraints, but instead a traversal. Calling `addTo()` on a `Variable` does not change the `Variable`'s state. (I just checked this). It only changes the `Traversal`'s state.